packages/url_launcher/url_launcher_ios/ios/Classes/URLLauncherPlugin.swift

```diff
@@ -0,0 +1,115 @@
+  }
+
+  private var currentSession: URLLaunchSession?
+  private let launcher: Launcher
+
+  init(launcher: Launcher = UIApplicationLauncher()) {
+    self.launcher = launcher
+  }
+
+  func canLaunchUrl(url: String) -> Bool {
+    guard let url = URL(string: url) else { return false }
+    return launcher.canOpenURL(url)
+  }
+
+  func launchUrl(
+    url: String, universalLinksOnly: Bool, completion: @escaping (Result<Bool, Error>) -> Void
+  ) {
+    guard let url = URL(string: url) else {
+      completion(Result.failure(invalidURLError(for: url)))
+      return
+    }
+    let options = [UIApplication.OpenExternalURLOptionsKey.universalLinksOnly: universalLinksOnly]
+    launcher.openURL(url, options: options) { success in
+      completion(Result.success(success))
+    }
+  }
+
+  func openUrlInSafariViewController(
+    url: String, completion: @escaping (Result<Bool, Error>) -> Void
+  ) {
+    guard let url = URL(string: url) else {
+      completion(Result.failure(invalidURLError(for: url)))
+      return
+    }
+
+    currentSession = URLLaunchSession(url: url, completion: completion)
+    guard let session = currentSession else { return }
+
+    session.didFinish = { [weak self] in
+      self?.currentSession = nil
+    }
+    topViewController?.present(session.safari, animated: true, completion: nil)
+  }
+
+  func closeSafariViewController() {
+    currentSession?.close()
+  }
+
+  var topViewController: UIViewController? {
+    // TODO(stuartmorgan) Provide a non-deprecated codepath. See
+    // https://github.com/flutter/flutter/issues/104117
+    UIApplication.shared.keyWindow?.rootViewController?.topViewController
+  }
+
+  /**
+    * Creates an error for an invalid URL string.
+    *
+    * @param url The invalid URL string
+    * @return The error to return
+    */
+  func invalidURLError(for url: String) -> Error {
+    GeneralError(
+      code: "argument_error", message: "Unable to parse URL", details: "Provided URL: \(url)")
+  }
+}
+
+/// This method recursively iterate through the view hierarchy
+/// to return the top most view controller.
+///
+/// It supports the following scenarios:
+///
+/// - The view controller is presenting another view.
+/// - The view controller is a UINavigationController.
+/// - The view controller is a UITabBarController.
+///
+/// @return The top most view controller.
+extension UIViewController {
+  var topViewController: UIViewController {
+    if let navigationController = self as? UINavigationController {
+      return navigationController.viewControllers.last?.topViewController ?? navigationController
+        .visibleViewController ?? navigationController
+    }
+    if let tabBarController = self as? UITabBarController {
+      return tabBarController.selectedViewController?.topViewController ?? tabBarController
+    }
+    if let presented = presentedViewController {
+      return presented.topViewController
+    }
+    return self
+  }
+}
+
+class GeneralError: Error {
```

Comment:
The information in this error class will be discarded by Pigeon.

@tarrinneal Can you look into https://github.com/flutter/flutter/issues/123851#issuecomment-1500567081 and see whether we can in fact use the extension approach? If not, we'll need to add Pigeon functionality before we could convert these errors this way.

Reply:
I don't think we support this kind of behavior in pigeon. But IIRC `FlutterError` should already be supported. 

I think I should update `FlutterError` in the engine so it conforms to `Error` automatically (e.g. to make it a subclass of `NSError`), but for now, you may have to write an empty extension here: 

```
extension FlutterError: Error {}
```

Reply:
Filed an issue here: https://github.com/flutter/flutter/issues/136081

Reply:
> but for now, you may have to write an empty extension here

The comment I linked to above is about validating whether we can do that without inter-plugin conflicts before we start shipping plugins that do it.

Reply:
Resolving since I checked, and it's safe to use the extension if we want to go that route.